check_diff/src/lib.rs

```diff
@@ -56,3 +162,146 @@ pub fn change_directory_to_path(dest: &Path) -> io::Result<()> {
     );
     return Ok(());
 }
+
+pub fn get_ld_lib_path() -> Result<String, CheckDiffError> {
+    let Ok(command) = Command::new("rustc").args(["--print", "sysroot"]).output() else {
+        return Err(CheckDiffError::FailedCommand(
+            "Error getting sysroot".to_string(),
+        ));
+    };
+
+    let Ok(sysroot) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting sysroot to string".to_string(),
+        ));
+    };
+
+    let ld_lib_path = format!("{}/lib", sysroot.trim_end());
+    return Ok(ld_lib_path);
+}
+
+pub fn get_cargo_version() -> Result<String, CheckDiffError> {
+    let Ok(command) = Command::new("cargo").args(["--version"]).output() else {
+        return Err(CheckDiffError::FailedCargoVersion(
+            "Failed to obtain cargo version".to_string(),
+        ));
+    };
+
+    let Ok(cargo_version) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting cargo version to string".to_string(),
+        ));
+    };
+
+    return Ok(cargo_version);
+}
+
+pub fn get_binary_version(
+    binary: &PathBuf,
+    ld_lib_path: &String,
+) -> Result<String, CheckDiffError> {
+    let binary_str = pathbuf_to_str(binary)?;
+    let Ok(command) = Command::new(binary_str)
+        .env("LD_LIB_PATH", ld_lib_path)
+        .args(["--version"])
+        .output()
+    else {
+        return Err(CheckDiffError::FailedVersioning(format!(
+            "Failed to get version for {:?}",
+            binary
+        )));
```

Comment:
`FailedVersioning` should probably just store an owned PathBuf and you can add this error message in the `Display` / `Debug` impl.